xds/internal/xdsclient/clientimpl_authority.go

```diff
@@ -121,28 +103,39 @@ func (c *clientImpl) newAuthorityLocked(config *bootstrap.ServerConfig) (_ *auth
 	}
 
 	// Make a new authority since there's no existing authority for this config.
-	nodeID := ""
-	if v3, ok := c.config.XDSServer.NodeProto.(*v3corepb.Node); ok {
-		nodeID = v3.GetId()
-	} else if v2, ok := c.config.XDSServer.NodeProto.(*v2corepb.Node); ok {
-		nodeID = v2.GetId()
-	}
-	ret := &authority{config: config, pubsub: pubsub.New(c.watchExpiryTimeout, nodeID, c.logger)}
-	defer func() {
-		if retErr != nil {
-			ret.close()
-		}
-	}()
-	ctr, err := newController(config, ret.pubsub, c.updateValidator, c.logger, nil)
+	ret, err := c.newAuthority(config)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to create new authority for config %q: %v", config.String(), err)
 	}
-	ret.controller = ctr
 	// Add it to the cache, so it will be reused.
 	c.authorities[configStr] = ret
 	return ret, nil
 }
 
+func (c *clientImpl) newAuthority(config *bootstrap.ServerConfig) (*authority, error) {
+	ret := &authority{
+		config:             config,
+		bootstrapCfg:       c.config,
+		serializer:         c.serializer,
+		resourceTypeGetter: c.resourceTypes.get,
+		watchExpiryTimeout: c.watchExpiryTimeout,
+		logger:             c.logger,
+		resources:          make(map[xdsresource.Type]map[string]*resourceState),
+	}
+	tr, err := transport.New(transport.Options{
+		ServerCfg:          *config,
+		UpdateHandler:      ret.handleResourceUpdate,
+		StreamErrorHandler: ret.newConnectionError,
+		Logger:             c.logger,
+	})
+	if err != nil {
+		ret.close()
```

Comment:
This is an awkward pattern for Go, speaking generally.  ("Construction of a type that contains sub-types that require initialization and could error and require shutdown".)  An interesting thing to consider and it's come up before in discussions I've had with menghanl.

`ret.close()` does nothing at all, but I can see having this for paranoia sake in case the initialization becomes more complex in the future.  But I'd also be happy to see this deleted and the `nil` check removed from `close`.  :shrug:

Reply:
Valid point.

Added a `newAuthority` function to create an authority. This will take care of creating the transport, and if that fails, it will return an error. And in this case, the caller of `newAuthority()` will not have to perform any cleanup.

Also, removed the `nil` check for the underlying transport from `authority.close()` since a valid `authority` now cannot exist without a valid `transport`.